File: guides/java/kora-java-guide-s3-app/src/main/java/ru/tinkoff/kora/guide/s3/service/UserService.java
package ru.tinkoff.kora.guide.s3.service;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.guide.s3.dto.UserRequest;
import ru.tinkoff.kora.guide.s3.dto.UserResponse;
import ru.tinkoff.kora.guide.s3.repository.UserRepository;
import ru.tinkoff.kora.http.server.common.HttpServerResponseException;

@Component
public final class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public UserResponse createUser(UserRequest request) {
        var generatedId = userRepository.save(request.name(), request.email());
        return new UserResponse(generatedId, request.name(), request.email(), LocalDateTime.now());
    }

    public Optional<UserResponse> getUser(String id) {
        return userRepository.findById(id);
    }

    public List<UserResponse> getUsers(int page, int size, String sort) {
        return userRepository.findAll().stream()
                .sorted(getComparator(sort))
                .skip((long) page * size)
                .limit(size)
                .toList();
    }

    public UserResponse updateUser(String id, UserRequest request) {
        boolean updated = userRepository.update(id, request.name(), request.email());
        if (!updated) {
            throw HttpServerResponseException.of(404, "User not found");
        }
        return new UserResponse(id, request.name(), request.email(), LocalDateTime.now());
    }

    public void deleteUser(String id) {
        boolean deleted = userRepository.deleteById(id);
        if (!deleted) {
            throw HttpServerResponseException.of(404, "User not found");
        }
    }

    private Comparator<UserResponse> getComparator(String sort) {
        return switch (sort.toLowerCase()) {
            case "name" -> Comparator.comparing(UserResponse::name);
            case "email" -> Comparator.comparing(UserResponse::email);
            case "createdat" -> Comparator.comparing(UserResponse::createdAt);
            default -> Comparator.comparing(UserResponse::name);
        };
    }
}

File: guides/kotlin/kora-kotlin-guide-openapi-http-server-app/src/main/kotlin/ru/tinkoff/kora/guide/openapi/httpserver/service/UserService.kt
package ru.tinkoff.kora.guide.openapi.httpserver.service

import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.guide.openapi.httpserver.dto.UserRequest
import ru.tinkoff.kora.guide.openapi.httpserver.dto.UserResponse
import ru.tinkoff.kora.guide.openapi.httpserver.repository.UserRepository
import java.time.LocalDateTime

@Component
class UserService(private val userRepository: UserRepository) {
    fun createUser(request: UserRequest): UserResponse {
        val id = userRepository.save(request.name, request.email)
        return UserResponse(id, request.name, request.email, LocalDateTime.now())
    }

    fun getUser(id: String): UserResponse? = userRepository.findById(id)
    fun getUsers(page: Int, size: Int, sort: String): List<UserResponse> = userRepository.findAll()
        .sortedWith(getComparator(sort))
        .drop(page * size)
        .take(size)

    fun updateUser(id: String, request: UserRequest): UserResponse {
        userRepository.update(id, request.name, request.email)
        return UserResponse(id, request.name, request.email, LocalDateTime.now())
    }

    fun deleteUser(id: String) {
        userRepository.deleteById(id)
    }

    private fun getComparator(sort: String): Comparator<UserResponse> = when (sort.lowercase()) {
        "name" -> compareBy { it.name }
        "email" -> compareBy { it.email }
        "createdat" -> compareBy { it.createdAt }
        else -> compareBy { it.name }
    }
}
